### ISSUE TITLE
TypeScript support?

### ISSUE THREAD

**Issue by JaKXz** [bcoe/nyc#176]
When scanning through the issues, there was a hint that theoretically, TypeScript could be possible to instrument and produce coverage for.

This is my first guess in my `package.json` for an Angular 2 project using karma:

```json
  "scripts": {
    "test": "npm run lint && nyc karma start",
    "lint": "tslint './src/**/*.ts'"
  },
  "nyc": {
    "extensions": [
      ".ts"
    ],
    "exclude": [
      "**/*.js",
      "**/*.test.ts",
      "**/tests.*.ts" //helper_files_like_'tests.entry.js'
    ]
  },
```

and I'm just getting an empty table.

**Comment by JaKXz** [comment]
I tried both `npm i -D nyc@latest` and `npm i -D bcoe/nyc#master` [at time of writing], because the regular alias was just getting me v5.6.0?

**Comment by bcoe** [comment]
@JaKXz good find, I should definitely promote `v6.0.0` to latest. You can install currently by running:

`npm i nyc@next`

Could you link me to an OSS project of yours that has `.ts` tests, I'd be happy to help figure things out.

**Comment by JaKXz** [comment]
@bcoe https://github.com/JaKXz/webpack-ts-demo/tree/chore/use-nyc is an old project, spent some time this morning updating it and cleaning up some errors. That branch has a pretty similar setup to what I've described here, with one `.spec.ts` file --  the `master` branch has the old `karma-coverage` setup, and instruments the transpiled code and so you do get coverage numbers, but they're skewed because of the extra boilerplate.

Feel free to email me if you need anything. Thanks for looking into this! :smile:

**Comment by bcoe** [comment]
@JaKXz any progress on this, hope things are going well.

**Comment by JaKXz** [comment]
@bcoe I haven't had any luck, no :( 

Was my project / branch installable without issues?

**Comment by JaKXz** [comment]
I did use `npm i nyc@next` in that branch :)

**Comment by JaKXz** [comment]
@bcoe I also tried here: https://github.com/JaKXz/angular2-redux-starter/commit/ec82374e628a55e566f5d015a8dbf1b078faf1e4 using `remap-istanbul`, following @novemberborn's [instructions](https://gist.github.com/novemberborn/e133e9edabbb380229ca) without success.

**Comment by novemberborn** [comment]
@JaKXz nyc can handle source maps out of the box now, provided the source maps are available when it instruments the source files. I'm not sure how well it would work with Karma though. In your branch you seem to be [hooking up Istanbul yourself](https://github.com/JaKXz/webpack-ts-demo/pull/3/files#diff-a2a3b7b0c9c3b4b93b4aebf4e3ec3cfbR37) which might conflict with nyc. And nyc might not work great with a Webpack bundle either.

**Comment by JaKXz** [comment]
@novemberborn good catch, but [removing the post-loader didn't work](https://github.com/JaKXz/webpack-ts-demo/pull/3) either.

My guess for sourcemaps was that they should be inline for `nyc` to read them, is that right?

**Comment by JaKXz** [comment]
More and more it just looks like the `.ts` extension is just not supported. No matter what configuration (with or without remap-istanbul and/or sourcemaps) I use, I don't get any files listed when I specify the `.ts` extension.

**Comment by bcoe** [comment]
@JaKXz I noticed a problem with the documentation digging into your problem, which I corrected `extensions` should have been documented as `extension`. Having said this, the approach you are using will not work with `nyc`:

You are using `ts-loader`, as far as I can determine this does not ultimately call a `require()` which is how nyc instruments files. As far as I can tell, you have a couple options:

1. you could try out this module which should allow you to instrument the files that are being loaded by web-pack: https://github.com/deepsweet/istanbul-instrumenter-loader
2. alternatively, you could test you typescript code in isolation, and use a module like this: https://github.com/theblacksmith/typescript-require

**Comment by JaKXz** [comment]
@bcoe thanks for the response. I fixed the typo in `extension` and have re-added `istanbul-instrumenter-loader` in my karma webpack config [as a postLoader], but the `nyc --reporter=json karma start` command is still returning an empty `coverage-final.json` object.

**Comment by bcoe** [comment]
@JaKXz `nyc` overrides `require()` and `child_process` to instrument code with Istanbul on the fly. It looks like the webpack is not compatible with this; rather than overriding the require statement it implements its own concept of loaders.

tldr; rather than using `nyc` just use `istanbul-instrumenter-loader`

1. make your loaders look something like this:

```js
//TODO use object.assign here
exports.typescript = {
  test: /\.ts$/,
  loaders: ['ng-annotate', 'istanbul-instrumenter', 'ts'],
  exclude: /\.spec\.ts$/
};

exports.typescriptTest = {
  test: /\.ts$/,
  loaders: ['ng-annotate', 'istanbul-instrumenter', 'ts']
};
```

2. make your `karma.conf.js` look something like this:

```js
reporters: ['mocha', 'coverage'],
```

3. take a look in your `/coverage` folder at the end of the day, and you're in flavor country:

<img width="1207" alt="screen shot 2016-03-14 at 11 03 28 pm" src="https://cloud.githubusercontent.com/assets/194609/13768985/22728a12-ea39-11e5-9db7-d66828e209c6.png">

PS., I'm flying in from Oakland and talking at Toronto JavaScript on the 22nd, if you happen to be around.

**Comment by JaKXz** [comment]
@bcoe thanks for the support! That is pretty much [as far as I'd gotten before](https://github.com/rangle/angular2-redux-starter/pull/44) but I was hoping that I could instrument the actual `.ts` files - I have to move on for now, but I will try again someday, perhaps with a fix to `remap-istanbul`.

Looking forward to meeting you! :smile:

**Comment by mohsen1** [comment]
I'll leave it here: [Setting up test coverage using Mocha, Istanbul, NYC with TypeScript](http://azimi.me/2016/09/30/nyc-mocha-typescript.html)

Thanks for amazing work! ❤️

**Comment by bcoe** [comment]
@mohsen1 this is wonderful, mind if I put a copy on our IstanbulJS website?

**Comment by mohsen1** [comment]
Not at all! I copied all of your code in my project! ;)

**Comment by seanpoulter** [comment]
FYI The link to @mohsen1's article has [changed](http://azimi.me/2016/09/30/nyc-mocha-typescript.1.html).

It's also possible to produce TypeScript coverage without ts-node by running tsc with source maps and letting Mocha use them with `--require source-map-support/register`. Nyc will map the rest for you.

**Comment by Izhaki** [comment]
I've managed to get this to work 

- with: `webpack`, `mocha`, `mocha-webpack`
- without: `babel`, `ts-node`, bells and whistles

# package.json

    "coverage": "nyc mocha-webpack",

# .nycrc

```json
{
    "include": [
      "src/**/*.ts"
    ],
    "extension": [
      ".ts"
    ],
    "reporter": [
      "html"
    ]
}
```

# mocha-webpack.opts

```
--webpack-config ./config/webpack/tests.js
--full-trace
src/**/*.spec.ts
```

# config/webpack/tests.js:

```javascript
var path = require('path');
var nodeExternals = require('webpack-node-externals');

var config = {
    target: 'node',

    externals: [ nodeExternals() ],

    output: {
        devtoolModuleFilenameTemplate: '[absolute-resource-path]',
        devtoolFallbackModuleFilenameTemplate: '[absolute-resource-path]?[hash]'
    },

    devtool: 'inline-source-map',

    resolve: {
        extensions: ['.ts', '.tsx', '.js', '.jsx']
    },

    module: {
        loaders: [{
            test: /\.ts?$/,
            loader: [ 'istanbul-instrumenter-loader', 'awesome-typescript-loader' ],
            exclude: /node_modules/
        }]
    }

};

module.exports = config;
```

**Comment by Nava2** [comment]
@seanpoulter you mentioned: 

> It's also possible to produce TypeScript coverage without ts-node by running tsc with source maps and letting Mocha use them with --require source-map-support/register.

Would this be possible to do while running mocha against `tsc`-compiled `js` files? If they have inline source maps. My scenario is that I have `tsc` compiling my library files to a folder, `lib-js`. All of my mocha tests than include the `lib-js` versions. 

I run coverage against the `js` files, but having the `ts` source maps respected would be excellent.

**Comment by bcoe** [comment]
@Nava2 @seanpoulter I recently created a [slack channel](http://devtoolscommunity.herokuapp.com/) for `nyc`, `istanbul`, `yargs`, etc., might be a good place to discuss some of these features.

**Comment by nikmash** [comment]
@seanpoulter were you able to get the all flag working so all files (even the ones that aren't required in your tests) will show up in the coverage report.

**Comment by seanpoulter** [comment]
I don't recall trying that @nikmash, sorry.

**Comment by nikmash** [comment]
So what I found was interesting is when using the `all` flag all the files that have been imported by the tests will show up as `.ts` files in the coverage report but all the files that have not been imported but are in the project will show up as `.js` files in the coverage report. Using `lcov` reporter.

**Comment by subhranshudas** [comment]
Can anybody provide the working configuration for JS + NYC + MOCHA-WEBPACK, but get a blank coverage report
```

var nodeExternals = require('webpack-node-externals');
var isCoverage = process.env.NODE_ENV === 'coverage';

module.exports = {
  output: {
    // use absolute paths in sourcemaps (important for debugging via IDE)
    devtoolModuleFilenameTemplate: '[absolute-resource-path]',
    devtoolFallbackModuleFilenameTemplate: '[absolute-resource-path]?[hash]'
  },
  module: {
    rules: [].concat(
      isCoverage ? {
          test: /\.(js)/,
          include: path.resolve('src'), // instrument only testing sources with Istanbul, after ts-loader runs
          loader: 'istanbul-instrumenter-loader'
      } : [],
      {
          test: /.js$/,
          exclude: /(node_modules|bower_components)/,
          loader: 'babel-loader',
      },
    ),
  },
  target: 'node',  // webpack should compile node compatible code
  externals: [nodeExternals()], // in order to ignore all modules in node_modules folder
  devtool: 'inline-cheap-module-source-map'
};
```
package.json

```
 "scripts": {
    "test": "./node_modules/mocha-webpack/bin/mocha-webpack ./tests/*.test.js",
    "coverage": "cross-env NODE_ENV=coverage nyc --reporter=lcov --reporter=text npm run test",
    "start": "webpack --watch",
    "build": "webpack"
  },```

can't see any files in src in the table.

**Comment by shirakaba** [comment]
> I'll leave it here: [Setting up test coverage using Mocha, Istanbul, NYC with TypeScript](http://azimi.me/2016/09/30/nyc-mocha-typescript.html)

This article was invaluable to me, but returned a 404; so here's the up-to-date URL:

[Setting up test coverage using Mocha, Istanbul, NYC with TypeScript](https://azimi.me/2016/09/30/nyc-mocha-typescript.1.html)

Note: The report was highlighted in some very nonsensical places when source maps were on, so I turned off `sourceMap` on `nyc` and removed `--require source-map-support/register` from the mocha options, which made for a more logical-looking coverage report. Not sure what's going on there.

**Comment by Cauany** [comment]
I'm having a problem with the configuration of the coverage of my project, where I'm using Vue.Js, Quasar framework and Typescript. Coverage reports are returning all files, but the files are not read in the coverage.

Dependencies -> nyc, @cypress-coverage, @istanbuljs/nyc-config-typescript, coveralls, istanbul-instrumenter-loader, istanbul-lib-coverage, @cypress/webpack-preprocessor and babel/preset-typescript.

Files
Babelrc ->  {"plugins": ["istanbul"],}
Nycrc.json -> {
  "nyc": {
    "extends": "@istanbuljs/nyc-config-typescript",
    "all": true
  }
}

Babel.config.js -> module.exports = {
  plugins: [
    ['babel-plugin-istanbul', {
      extension: ['.js', '.vue','.ts']
    }]
  ]
};

![Screenshot_20201109_165001](https://user-images.githubusercontent.com/61061582/98589197-b2d53000-22ab-11eb-8941-5b0c1e27242e.png)
